README.md
# Purpose

[![MELPA](http://melpa.org/packages/window-purpose-badge.svg)](http://melpa.org/#/window-purpose)
[![MELPA Stable](http://stable.melpa.org/packages/window-purpose-badge.svg)](http://stable.melpa.org/#/window-purpose)
[![Build Status](https://travis-ci.org/bmag/emacs-purpose.svg?branch=master)](https://travis-ci.org/bmag/emacs-purpose)
[![Coverage Status](https://coveralls.io/repos/bmag/emacs-purpose/badge.svg?branch=master)](https://coveralls.io/r/bmag/emacs-purpose?branch=master)

**Note: a full explanation can be found in the [GitHub wiki](https://github.com/bmag/emacs-purpose/wiki).**

## Introduction

Purpose ("window-purpose" on MELPA) provides a new window management system for
Emacs, which gives you a better control over where Emacs displays buffers.

With Purpose, each buffer has a configurable "purpose" and each window
can interactively be dedicated to a certain "purpose". When you dedicate
a window (<kbd>C-c , d</kbd>), Purpose makes sure that this window will be used
only for buffers which have the same purpose as the buffer that is
currently displayed in that window. The purpose of a buffer can be
customized via the variables `purpose-user-mode-purposes`,
`purpose-user-name-purposes`, `purpose-user-regexp-purposes` and
`purpose-use-default-configuration` (see the
[wiki](https://github.com/bmag/emacs-purpose/wiki/Purpose-Configuration)).

### Supported Emacs Versions
Purpose is tested with Emacs versions 24.3, 24.4 and 24.5. Backward-compatibility for earlier versions might be added in the future - if you want such support, contact me.

## Quickstart

### Activate Purpose
Manually: <kbd>M-x purpose-mode</kbd>

In your init file:
```elisp
(require 'window-purpose)
(purpose-mode)
```

### Configure Purpose
Manually: <kbd>M-x customize-group purpose</kbd>. Look at:
- "Purpose User Mode Purposes": recognize purpose according to major mode
- "Purpose User Name Purposes": recognize purpose according to buffer
  name (for exact names)
- "Purpose User Regexp Purposes": recognize purpose according to buffer
  name (for name patterns)
- "Purpose Use Default Configuration": toggle default configuration
  on/off

In init file:
```elisp
(add-to-list 'purpose-user-mode-purposes '(<major-mode> . <purpose>))
(add-to-list 'purpose-user-name-purposes '(<name> . <purpose>))
(add-to-list 'purpose-user-regexp-purposes '(<pattern> . <purpose>))
(setq purpose-use-default-configuration t) ; not really necessary, default is t
(purpose-compile-user-configuration) ; activates your changes
```

### Useful Commands
| Key         | Command                                                                                                                   |
| :---------- | :------------------------------------------------------------------------------------------------------------------------ |
| <kbd>C-c , b</kbd>   | `purpose-switch-buffer-with-purpose`: switch to a buffer with the same purpose as the current one                         |
| <kbd>C-u C-x b</kbd> | `switch-buffer-without-purpose`: switch to a buffer, but don't use Purpose for it. Handy for changing the current layout. |
| <kbd>C-c , d</kbd>   | `purpose-toggle-window-purpose-dedicated`                                                                                 |
| <kbd>C-c , D</kbd>   | `purpose-toggle-window-buffer-dedicated`                                                                                  |
| <kbd>C-c , 1</kbd>   | `purpose-delete-non-dedicated-windows`                                                                                    |
|             | `purpose-save-window-layout`: save current layout (by name)                                                                |
|             | `purpose-load-window-layout`: load layout (by name)                                                                       |
|             | `purpose-save-window-layout-file`: save current layout directly to file                                                                |
|             | `purpose-load-window-layout-file`: load layout directly from file                                                                  |
|             | `purpose-reset-window-layout`: reload previously loaded layout                                                            |

### Example: Simple Python Layout
How to get a simple and persistent layout for coding in Python that
looks like this:

![simple python layout](https://github.com/bmag/emacs-purpose/blob/master/images/simple-python-layout.png)

#### step 1: configuration
```elisp
(add-to-list 'purpose-user-mode-purposes '(python-mode . py))
(add-to-list 'purpose-user-mode-purposes '(inferior-python-mode . py-repl))
(purpose-compile-user-configuration)
```

#### step 2: change window layout
If you have a previously saved layout, you can load it with
`purpose-load-window-layout` and skip step 2 and step 3.

1. open a Python file
2. <kbd>C-c , d</kbd> (`purpose-toggle-window-purpose-dedicated`) so window is
   dedicated ("[py]" in the status bar will change to "[py!]")
3. <kbd>C-x 1</kbd> (`delete-other-windows`)
4. <kbd>C-x 2</kbd> (`split-window-below`)
5. <kbd>C-c C-z</kbd> (`python-shell-switch-to-shell`)
6. <kbd>C-c , d</kbd> so window is dedicated
7. <kbd>C-x o</kbd> (`other-window`) to select the python file's window
8. <kbd>C-x ^</kbd> (`enlarge-window`) until you like the sizes of the windows

#### step 3: save window layout
<kbd>M-x purpose-save-window-layout</kbd>

## Using Purpose

### Dedicating windows
Dedicating a window limits which buffers will be displayed in it. There are two
types of window dedication: buffer-dedication and purpose-dedication.

Use `purpose-toggle-window-buffer-dedicated` to dedicate a window to its buffer.
This window will not display any other buffer while it is buffer-dedicated. A
"#" in the mode-line next to the window's purpose indicates that the window is
buffer-dedicated.

Use `purpose-toggle-window-purpose-dedicated` to dedicate a window to its
purpose. This window will only display buffers with the same purpose. A "!" in
the mode-line next to the window's purpose indicates that the window is
purpose-dedicated.

You can delete all non-dedicated windows by using
`purpose-delete-non-dedicated-windows`.

### Switching buffers
When switching buffers, Purpose will display the new buffer in the correct
window, according to the current configuration.

Use `switch-to-buffer` to switch to any buffer. The buffer will be displayed
according to the current purpose-configuration.

Use `purpose-switch-buffer-with-purpose` to switch to another buffer with the
same purpose as the current buffer.

Use `purpose-switch-buffer-with-some-purpose` to select a purpose and then
switch to a buffer with that purpose.

Use `switch-buffer-without-purpose` to switch to any buffer. The buffer will be
displayed using Emacs' original behavior. This is useful when you want to change
the window layout.

Use `purpose-set-window-purpose` to switch the purpose of the current window. If
there is a buffer with the chosen purpose, that buffer will be displayed in the
current window. Otherwise, a dummy buffer will be created and used.

### Changing layout
Purpose lets you save, load and reset the window layout.

Use `purpose-save-window-layout` to save the current window layout. The layout
will be saved in a directory of your choice, in a file named
`<layout-name>.window-layout`.

Use `purpose-load-window-layout` to load a window layout. The available layouts
are located the directories specified in customizable variable `purpose-layout-dirs`.

Use `purpose-save-window-layout-file` to save the current window layout directly to
a file of your choice.

Use `purpose-load-window-layout-file` to load a window layout directly from a file
of your choice.

Use `purpose-reset-window-layout` to reset the window layout to the latest
layout that you loaded.

In addition to window layouts, Purpose also lets save, load and reset the frame
layout. A "frame layout" consists of the window layouts of multiple frames.
All of the window layout commands have frame layout equivalents, e.g.
`purpose-load-frame-layout` is equivalent to `purpose-load-window-layout`.

## Customizing Purpose

### Prompts
In the past, the variable `purpose-preferred-prompt`, allowed you to choose whether you want Purpose to use
IDO or Helm when it needs information from the user. This variable has been removed. Instead, the prompt
relies on the normal mechanism. For Helm prompts enable `helm-mode`, for Ivy mode enable `ivy-mode`, for
IDO prompts enable `ido-mode`, etc. If using `ido-mode`, note that because how IDO works it doesn't
cover all commands, and some commands won't have an IDO prompt. To work around that, please see the
package [ido-completion-read+](https://github.com/DarwinAwardWinner/ido-completing-read-plus) (formerly
ido-ubiquitous).

### Changing purpose configuration
Purpose lets you define your own purposes. You can do so by using the variables
`purpose-user-mode-purposes`, `purpose-user-name-purposes` and
`purpose-user-regexp-purposes`. You can also deactivate the default purpose
configuration if it bothers you, by setting `purpose-use-default-configuration`
to nil.

### Changing display rules
If you want, you can the rules of how certain buffers are displayed. To do so,
use the variable `purpose-special-action-sequences`. Let's explain this with an
example. The following code makes all help buffers appear in a separate
frame. This means you will get a "popup" frame for help buffers.

```elisp
(setq pop-up-frames t) ; allows emacs to popup new frames
;; give help buffers the 'popup-frame purpose
(add-to-list 'purpose-user-mode-purposes
             '(help-mode . popup-frame))
(purpose-compile-user-configuration)
;; new rules for buffers with the 'popup-frame purpose
(add-to-list 'purpose-special-action-sequences
             '(popup-frame
               purpose-display-reuse-window-buffer
               purpose-display-reuse-window-purpose
               purpose-display-pop-up-frame))
```

### Respect purposes when killing a buffer
When killing a visible buffer, Emacs has to decide which buffer to show instead.
Enabling the `purpose-x-kill` extension will make Emacs consider the purpose of
the window that needs to show a new buffer. If the window is purpose-dedicated,
the killed buffer is replaced with another buffer with the same purpose. If there
are no buffers with the same purpose as the killed buffer, the window is deleted.
To enable the `purpose-x-kill` extension:

```elisp
(require 'window-purpose-x)
(purpose-x-kill-setup)
```

### Location of Layout Files
Window layout and frame layout files are stored in the directories specified by
`purpose-layout-dirs`. By default, its value is `("~/.emacs.d/layouts/")`. To store
layouts in a different location, simply change the value of this variable.

## Using Purpose with other packages
See [Integration With Other Packages](https://github.com/bmag/emacs-purpose/wiki/Integration-With-Other-Packages)
for information about how some packages relate to Purpose.


CHANGELOG.org
#+TITLE: Changelog

* Not Part of a Tagged Release Yet
* Release 1.8
** Breaking Changes
*** [[https://github.com/bmag/emacs-purpose/issues/144][#144]]: purpose-x-magit extension: replace ~magit~ purpose with ~Magit~
    When using the purpose-x-magit extension via
    ~purpose-x-magit-single-on~, the extension no longer defines a
    ~magit~ purpose. Instead it defines an identical purpose named
    ~Magit~. This way, the purpose can be used as a condition in
    ~purpose-special-action-sequences~.

    Those who use purpoxe-x-magit via ~purpose-x-magit-multi-on~ are
    not affected by this change.
** Enhancements
*** [[https://github.com/bmag/emacs-purpose/pull/172][#172]]: Rework Travis CI tests configuration and add Emacsen up to 27.1
*** [[https://github.com/bmag/emacs-purpose/issues/149][#149]], [[https://github.com/bmag/emacs-purpose/pull/152][#152]]: Lower memeory usage when calculating a buffer's purpose
** Bug Fixes
*** [[https://github.com/bmag/emacs-purpose/pull/182][#182]]: Prevent ~debug~ command from bypassing purpose
*** [[https://github.com/bmag/emacs-purpose/issues/166][#166]], [[https://github.com/bmag/emacs-purpose/pull/180][#180]]: Toggle integration fixes when ~purpose-mode~ is toggled
Previously, the integration fixes would remain active even when ~purpose-mode~
was disabled, disrupting the normal display behavior of integrated packages and
features.
*** [[https://github.com/bmag/emacs-purpose/issues/153][#153]], [[https://github.com/bmag/emacs-purpose/pull/177][#177]]: purpose-x-popwin extension: quit-window closes popup window when no previous buffers can be shown in it
*** [[https://github.com/bmag/emacs-purpose/pull/171][#171]]: Reimplement ~whitespace-display-window~ using purposes
Original ~whitespace-display-window~ doesn't go through ~display-buffer~, thus
disregarding ~window-purpose~ entirely.
*** [[https://github.com/bmag/emacs-purpose/pull/170][#170]]: Reimplement ~edebug-pop-to-buffer~ using purposes
Original ~edebug-pop-to-buffer~ doesn't go through ~display-buffer~, thus
disregarding ~window-purpose~ entirely.
*** [[https://github.com/bmag/emacs-purpose/pull/169][#169]]: Prevent Isearch help commands from bypassing purpose
*** [[https://github.com/bmag/emacs-purpose/pull/168][#168]]: purpose-x-magit extension: prevent ~magit-display-buffer-function~ from bypassing purpose
*** [[https://github.com/bmag/emacs-purpose/pull/167][#167]]: Prevent ~next-error~ from bypassing purpose
*** [[https://github.com/bmag/emacs-purpose/pull/166][#166]]: Properly format defcustom package-version
*** [[https://github.com/bmag/emacs-purpose/issues/133][#133]]: Neotree fix honors ~neo-window-position~
*** [[https://github.com/bmag/emacs-purpose/issues/137][#137]]: ~purpose-change-buffer~ to ignore ~dedicated~ argument in new Emacsen
    As of Emacs 27.0.50, the function ~window--display-buffer~ no longer accepts
    a ~dedicated~ argument, and instead relies on ~alist~ and
    ~display-buffer-mark-dedicated~ to decide dedication state of the window. As
    result, ~purpose-change-buffer~ ignores the ~dedicated~ argument when using
    Emacs version 27 or newer. Older Emacs versions are not affected.
*** [[https://github.com/bmag/emacs-purpose/pull/134][#134]]: add support for Zone
*** [[https://github.com/bmag/emacs-purpose/issues/78#issuecomment-467664376][#78 (comment)]] purpose-x-persp extension: ~persp-curr~ is now a function 
* Release 1.7
** Breaking Changes
*** removed support for Emacs 24.3
    Supported versions are Emacs 24.4 and newer (up to 26.1 at the
    time of writing). The branch ~release-1.6~ will continue to
    support Emacs 24.3 and is available for those who need it.
** Bug Fixes
*** [[https://github.com/bmag/emacs-purpose/pull/116][#116]]: honor ~display-buffer-mark-dedicated~
* Release 1.6.1
** Breaking Changes
*** [[https://github.com/bmag/emacs-purpose/issues/84][#84]], [[https://github.com/bmag/emacs-purpose/pull/125][#125]]:  remove option ~purpose-preferred-prompt~ and prompts shims.
    Removed ~completing-read~ and ~read-file-name~ shims, thus the
    custom variable ~purpose-preferred-prompt~ has been removed. From the README:
    #+BEGIN_QUOTE
    In the past, the variable `purpose-preferred-prompt`, allowed you
    to choose whether you want Purpose to use IDO or Helm when it
    needs information from the user. This variable has been
    removed. Instead, the prompt relies on the normal mechanism. For
    Helm prompts enable `helm-mode`, for Ivy mode enable `ivy-mode`,
    for IDO prompts enable `ido-mode`, etc. If using `ido-mode`, note
    that because how IDO works it doesn't cover all commands, and some
    commands won't have an IDO prompt. To work around that, please see
    the package
    [ido-completion-read+](https://github.com/DarwinAwardWinner/ido-completing-read-plus)
    (formerly ido-ubiquitous).
    #+END_QUOTE
** Enhancements
*** [[https://github.com/bmag/emacs-purpose/issues/113][#113]], [[https://github.com/bmag/emacs-purpose/pull/114][#114]]: default to ~edit~ purpose for buffers visiting files
    If a buffer is visiting a file, its default purpose (if no other
    purpose matches) will be ~edit~.
*** Emacsen 25.1, 25.3 and 26.1 added to Travis CI tests
    Emacs 25.2 not added, because it is almost exactly the same as
    25.3, and the difference doesn't affect Purpose's behavior.
** Bug Fixes
*** added some missing default purposes for some modes
    - css-mode: edit
    - eshell-mode, term-mode: terminal
*** [[https://github.com/bmag/emacs-purpose/issues/110][#110]]: wrong window layout after using magit-popup's help system
    Added a fix in ~window-purpose-fixes.el~.
*** [[https://github.com/bmag/emacs-purpose/issues/111][#111]]: wrong layout when using org-goto
    Added a fix in ~window-purpose-fixes.el~.
*** [[https://github.com/bmag/emacs-purpose/pull/123][#123]] purpose-x-popwin: fix custom type mismatches

window-purpose.el
;;; window-purpose.el --- Purpose-based window management for Emacs -*- lexical-binding: t -*-

;; Copyright (C) 2015-2021 Bar Magal & contributors

;; Author: Bar Magal
;; Package: purpose
;; Version: 1.8
;; Keywords: frames
;; Homepage: https://github.com/bmag/emacs-purpose
;; Package-Requires: ((emacs "24.4") (let-alist "1.0.3") (imenu-list "0.1"))

;; This file is not part of GNU Emacs.

;; This program is free software: you can redistribute it and/or modify
;; it under the terms of the GNU General Public License as published by
;; the Free Software Foundation, either version 3 of the License, or
;; (at your option) any later version.

;; This program is distributed in the hope that it will be useful,
;; but WITHOUT ANY WARRANTY; without even the implied warranty of
;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
;; GNU General Public License for more details.

;; You should have received a copy of the GNU General Public License
;; along with this program.  If not, see <http://www.gnu.org/licenses/>.

;;; Commentary:

;; ---------------------------------------------------------------------
;; Full information can be found on GitHub:
;; https://github.com/bmag/emacs-purpose/wiki
;; ---------------------------------------------------------------------

;; Purpose is a package that introduces the concept of a "purpose" for
;; windows and buffers, and then helps you maintain a robust window
;; layout easily.

;; Installation and Setup:
;; Install Purpose from MELPA, or download it manually from GitHub. If
;; you download manually, add these lines to your init file:
;;    (add-to-list 'load-path "/path/to/purpose")
;;    (require 'window-purpose)
;; To activate Purpose at start-up, add this line to your init file:
;;    (purpose-mode)

;; Purpose Configuration:
;; Customize `purpose-user-mode-purposes', `purpose-user-name-purposes',
;; `purpose-user-regexp-purposes' and
;; `purpose-use-default-configuration'.

;; Basic Usage:
;; 1. Load/Save window/frame layout (see `purpose-load-window-layout',
;;    `purpose-save-window-layout', etc.)
;; 2. Use regular switch-buffer functions - they will not mess your
;;    window layout (Purpose overrides them).
;; 3. If you don't want a window's purpose/buffer to change, dedicate
;;    the window:
;;    C-c , d: `purpose-toggle-window-purpose-dedicated'
;;    C-c , D: `purpose-toggle-window-buffer-dedicated'
;; 4. To use a switch-buffer function that ignores Purpose, prefix it
;;    with C-u. For example, [C-u C-x b] calls
;;    `switch-buffer-without-purpose'.

;;; Code:

(require 'easymenu)
(require 'window-purpose-utils)
(require 'window-purpose-configuration)
(require 'window-purpose-core)
(require 'window-purpose-layout)
(require 'window-purpose-switch)
(require 'window-purpose-prefix-overload)
(require 'window-purpose-fixes)

(defconst purpose-version "1.8.0"
  "Purpose's version.")



;;; Commands that work with both `ido' and `helm'
;; `helm' doesn't work well with `ido-find-file' (and other `ido' commands),
;; but `find-file' can't benefit from all of `ido''s features, so we create
;; commands that know when to use `ido-find-file' and when to use `find-file'.
;; the result: Purpose works well both with `ido' and `helm'!!!

(defmacro purpose-ido-caller (ido-fn other-fn)
  "Create an interactive lambda to conditionally call an ido command.
The lambda calls IDO-FN interactively when `ido-mode' is on, otherwise
it calls OTHER-FN interactively.
Example:
  (purpose-ido-caller #'ido-find-file #'find-file)"
  (declare (indent nil) (debug (function-form function-form)))
  `(lambda (&rest _args)
     (interactive)
     (call-interactively (if ido-mode ,ido-fn ,other-fn))))

(defalias 'purpose-friendly-find-file
  (purpose-ido-caller #'ido-find-file #'find-file)
  "Call `find-file' or `ido-find-file' intelligently.
If `ido-mode' is on, call `ido-find-file'.  Otherwise, call `find-file'.
This allows Purpose to work well with both `ido' and `helm'.")

(defalias 'purpose-friendly-find-file-other-window
  (purpose-ido-caller #'ido-find-file-other-window #'find-file-other-window)
  "Call `find-file-other-window' or `ido-find-file-other-window'
intelligently.
If `ido-mode' is on, call `ido-find-file-other-window'.  Otherwise, call
`find-file-other-window'.
This allows Purpose to work well with both `ido' and `helm'.")

(defalias 'purpose-friendly-find-file-other-frame
  (purpose-ido-caller #'ido-find-file-other-frame #'find-file-other-frame)
  "Call `find-file-other-frame' or `ido-find-file-other-frame'
intelligently.
If `ido-mode' is on, call `ido-find-file-other-frame'.  Otherwise, call
`find-file-other-frame'.
This allows Purpose to work well with both `ido' and `helm'.")

(defalias 'purpose-friendly-switch-buffer
  (purpose-ido-caller #'ido-switch-buffer #'switch-to-buffer)
  "Call `switch-to-buffer' or `ido-switch-buffer' intelligently.
If `ido-mode' is on, call `ido-switch-buffer'.  Otherwise, call
`switch-to-buffer'.
This allows Purpose to work well with both `ido' and `helm'.")

(defalias 'purpose-friendly-switch-buffer-other-window
  (purpose-ido-caller #'ido-switch-buffer-other-window
                      #'switch-to-buffer-other-window)
  "Call `switch-to-buffer-other-window' or
`ido-switch-buffer-other-window' intelligently.
If `ido-mode' is on, call `ido-switch-buffer-other-window'.  Otherwise,
call `switch-to-buffer-other-window'.
This allows Purpose to work well with both `ido' and `helm'.")

(defalias 'purpose-friendly-switch-buffer-other-frame
  (purpose-ido-caller #'ido-switch-buffer-other-frame
                      #'switch-to-buffer-other-frame)
  "Call `switch-to-buffer-other-frame' or
`ido-switch-buffer-other-frame' intelligently.
If `ido-mode' is on, call `ido-switch-buffer-other-frame'.  Otherwise,
call `switch-to-buffer-other-frame'.
This allows Purpose to work well with both `ido' and `helm'.")



;;; Commands for using Purpose-less behavior
(defalias 'find-file-without-purpose
  (without-purpose-command #'find-file))

(defalias 'find-file-other-window-without-purpose
  (without-purpose-command #'find-file-other-window))

(defalias 'find-file-other-frame-without-purpose
  (without-purpose-command #'find-file-other-frame))

(defalias 'switch-buffer-without-purpose
  (without-purpose-command #'switch-to-buffer))

(defalias 'switch-buffer-other-window-without-purpose
  (without-purpose-command #'switch-to-buffer-other-window))

(defalias 'switch-buffer-other-frame-without-purpose
  (without-purpose-command #'switch-to-buffer-other-frame))


;;; Overloaded commands: (C-u to get original Purpose-less behavior)
(define-purpose-prefix-overload purpose-find-file-overload
  '(purpose-friendly-find-file find-file-without-purpose))

(define-purpose-prefix-overload purpose-find-file-other-window-overload
  '(purpose-friendly-find-file-other-window find-file-other-window-without-purpose))

(define-purpose-prefix-overload purpose-find-file-other-frame-overload
  '(purpose-friendly-find-file-other-frame find-file-other-frame-without-purpose))

(define-purpose-prefix-overload purpose-switch-buffer-overload
  '(purpose-friendly-switch-buffer
    switch-buffer-without-purpose
    purpose-switch-buffer-with-purpose))

(define-purpose-prefix-overload purpose-switch-buffer-other-window-overload
  '(purpose-friendly-switch-buffer-other-window
    switch-buffer-other-window-without-purpose
    purpose-switch-buffer-with-purpose-other-window))

(define-purpose-prefix-overload purpose-switch-buffer-other-frame-overload
  '(purpose-friendly-switch-buffer-other-frame
    switch-buffer-other-frame-without-purpose
    purpose-switch-buffer-with-purpose-other-frame))



(define-purpose-prefix-overload purpose-delete-window-at
  '(purpose-delete-window-at-bottom
    purpose-delete-window-at-right
    purpose-delete-window-at-top
    purpose-delete-window-at-left))

(defvar purpose-mode-map
  (let ((map (make-sparse-keymap)))
    (define-key map (kbd "C-x C-f") #'purpose-find-file-overload)
    (define-key map (kbd "C-x 4 f") #'purpose-find-file-other-window-overload)
    (define-key map (kbd "C-x 4 C-f") #'purpose-find-file-other-window-overload)
    (define-key map (kbd "C-x 5 f") #'purpose-find-file-other-frame-overload)
    (define-key map (kbd "C-x 5 C-f") #'purpose-find-file-other-frame-overload)

    (define-key map (kbd "C-x b") #'purpose-switch-buffer-overload)
    (define-key map (kbd "C-x 4 b") #'purpose-switch-buffer-other-window-overload)
    (define-key map (kbd "C-x 5 b") #'purpose-switch-buffer-other-frame-overload)

    ;; Helpful for quitting temporary windows. Close in meaning to
    ;; `kill-buffer', so we map it to a close key ("C-x j" is close to
    ;; "C-x k")
    (define-key map (kbd "C-x j") #'quit-window)

    ;; We use "C-c ," for compatibility with key-binding conventions
    (define-key map (kbd "C-c ,") 'purpose-mode-prefix-map)
    (define-prefix-command 'purpose-mode-prefix-map)
    (cl-loop for (key . def)
             in '(;; switch to any buffer
                  ("o" . purpose-switch-buffer)
                  ("[" . purpose-switch-buffer-other-frame)
                  ("p" . purpose-switch-buffer-other-window)
                  ;; switch to buffer with current buffer's purpose
                  ("b" . purpose-switch-buffer-with-purpose)
                  ("4 b" . purpose-switch-buffer-with-purpose-other-window)
                  ("5 b" . purpose-switch-buffer-with-purpose-other-frame)
                  ;; toggle window dedication (buffer, purpose)
                  ("d" . purpose-toggle-window-purpose-dedicated)
                  ("D" . purpose-toggle-window-buffer-dedicated)
                  ;; delete windows
                  ("w" . purpose-delete-window-at)
                  ("1" . purpose-delete-non-dedicated-windows))
             do (define-key purpose-mode-prefix-map key def))
    map)
  "Keymap for Purpose mode.")

(easy-menu-define purpose-menu purpose-mode-map "Purpose Mode"
  '("Purpose"
    ["Toggle Buffer Dedication" purpose-toggle-window-buffer-dedicated
     :help "Toggle current window's dedication to its current buffer"]
    ["Toggle Purpose Dedication" purpose-toggle-window-purpose-dedicated
     :help "Toggle current window's dedication to its current purpose"]
    ["Change Window Purpose" purpose-set-window-purpose
     :help "Select a purpose for the current window and change its buffer accordingly"]
    ["Delete Non-Dedicated Windows" purpose-delete-non-dedicated-windows
     :help "Delete all windows that aren't dedicated to their buffer or purpose"]
    ["Load Window Layout" purpose-load-window-layout t]
    ["Save Window Layout" purpose-save-window-layout t]
    ["Load Frame Layout" purpose-load-frame-layout t]
    ["Save Frame Layout" purpose-save-frame-layout t]))

;; https://www.gnu.org/software/emacs/manual/html_node/elisp/Menu-Keymaps.html#Menu-Keymaps
;; (defvar purpose-menu-bar-map (make-sparse-keymap "Purpose"))

(defun purpose--modeline-string ()
  "Return the presentation of a window's purpose for display in the
modeline.  The basic form of the string is \"[<purpose>]\".  If the
window is purpose-dedicated, add a \"!\" before \"]\".  If the window is
buffer-dedicated, add a \"#\" before \"]\".
Some examples:
\"[edit]\": window's purpose is 'edit, and it is not dedicated.
\"[edit!]\": window is dedicated to 'edit purpose.
\"[edit#]\": window's purpose is 'edit, and it is dedicated to its
           current buffer.
\"[edit!#]\": window is dedicated to 'edit purpose and to its current buffer."
  (format " [%s%s%s]"
          (purpose-window-purpose)
          (if (purpose-window-purpose-dedicated-p) "!" "")
          (if (window-dedicated-p) "#" "")))

(defun purpose--add-advices ()
  "Add all advices needed for Purpose to work.
This function is called when `purpose-mode' is activated."
  (advice-add 'switch-to-buffer :around #'purpose-switch-to-buffer-advice)
  (advice-add 'switch-to-buffer-other-window :around #'purpose-switch-to-buffer-other-window-advice)
  (advice-add 'switch-to-buffer-other-frame :around #'purpose-switch-to-buffer-other-frame-advice)
  (advice-add 'pop-to-buffer :around #'purpose-pop-to-buffer-advice)
  (advice-add 'pop-to-buffer-same-window :around #'purpose-pop-to-buffer-same-window-advice)
  (advice-add 'display-buffer :around #'purpose-display-buffer-advice))

(defun purpose--remove-advices ()
  "Remove all advices needed for Purpose to work.
This function is called when `purpose-mode' is deactivated."
  (advice-remove 'switch-to-buffer #'purpose-switch-to-buffer-advice)
  (advice-remove 'switch-to-buffer-other-window #'purpose-switch-to-buffer-other-window-advice)
  (advice-remove 'switch-to-buffer-other-frame #'purpose-switch-to-buffer-other-frame-advice)
  (advice-remove 'pop-to-buffer #'purpose-pop-to-buffer-advice)
  (advice-remove 'pop-to-buffer-same-window #'purpose-pop-to-buffer-same-window-advice)
  (advice-remove 'display-buffer #'purpose-display-buffer-advice))

;;;###autoload
(define-minor-mode purpose-mode nil
  :global t :lighter (:eval (purpose--modeline-string))
  (if purpose-mode
      (progn
        (purpose--add-advices)
        (setq display-buffer-overriding-action
              '(purpose--action-function . nil))
        (setq purpose--active-p t)
        (unless purpose-fix-togglers-hook
          (purpose-fix-install))
        (run-hooks 'purpose-fix-togglers-hook))

    (purpose--remove-advices)
    (setq purpose--active-p nil)
    (run-hooks 'purpose-fix-togglers-hook)))

(push '(purpose-dedicated . writable) window-persistent-parameters)
(provide 'window-purpose)
;;; window-purpose.el ends here
